core/src/cache.rs

```diff
@@ -1037,36 +1037,40 @@ impl CacheHub {
         Ok(())
     }
 
-    /// Applies a custom transform to an input and its imports, leaving them in the same state as
-    /// before. Requires that the input has been parsed. In order for the transform to apply to
-    /// imports, they need to have been resolved.
+    /// Applies a custom transform to an input and its imports. Requires that source has been typechecked.
+    ///
+    /// If multiple invocations of `custom_transform` are needed, you must supply `transform_id` with
+    /// with a number higher than that of all previous invocations.
     pub fn custom_transform<E>(
         &mut self,
         file_id: FileId,
+        transform_id: usize,
         f: &mut impl FnMut(&mut CacheHub, RichTerm) -> Result<RichTerm, E>,
     ) -> Result<(), CacheError<E>> {
         match self.terms.entry_state(file_id) {
-            Some(state) if state >= EntryState::Parsed => {
-                if state < EntryState::Transforming {
+            Some(state) if state >= EntryState::Typechecked => {
+                if state.needs_custom_transform(transform_id) {
```

Comment:
I think this can just be `state < EntryState::CustomTransformed { transform_id }`. That way it looks like all our other state comparisons, where we're using the order to figure out what's been done.

Reply:
Done!

Reply:
Ah, it's a gotcha I didn't know about, good catch @twitchyliquid64.